compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -198,9 +198,27 @@ fn check_static_inhabited(tcx: TyCtxt<'_>, def_id: LocalDefId) {
 /// projections that would result in "inheriting lifetimes".
 fn check_opaque(tcx: TyCtxt<'_>, id: hir::ItemId) {
     let item = tcx.hir().item(id);
-    let hir::ItemKind::OpaqueTy(hir::OpaqueTy { origin, .. }) = item.kind else {
-        tcx.sess.delay_span_bug(item.span, "expected opaque item");
-        return;
+    let origin = match item.kind {
+        hir::ItemKind::OpaqueTy(hir::OpaqueTy { origin, in_trait: true, .. }) => {
+            let trait_fn = tcx.impl_trait_in_trait_parent_fn(id.owner_id.to_def_id());
+            // Only check the validity of this opaque type if the function has a default body
+            if let hir::Node::TraitItem(hir::TraitItem {
+                kind: hir::TraitItemKind::Fn(_, hir::TraitFn::Provided(_)),
+                ..
+            }) = tcx.hir().get_by_def_id(trait_fn.expect_local())
+            {
+                origin
+            } else {
+                return;
```

Comment:
Why is this needed now?

Reply:
Because we want to avoid checking the validity of the opaque if the trait has no default body.

Reply:
Of course. But we had opaque types before in trait methods with no default body. What *changed* to make it so that this code path is being called now?

Reply:
Ok, I see your point. Maybe it's more obvious if we do at this level as it's done in `ImplTraitPlaceholder` https://github.com/spastorino/rust/blob/e352207f21c8524f68f4dedb3ee7c55cf392b734/compiler/rustc_hir_analysis/src/check/check.rs#L587-L595

Reply:
That link is broken, but yeah, I think I agree.

Reply:
https://github.com/spastorino/rust/blob/master/compiler/rustc_hir_analysis/src/check/check.rs#L588-L598